plugins/datastreamer_plugin/fastdds/Participant.cpp

```diff
@@ -0,0 +1,456 @@
+        eprosima::fastdds::dds::DomainParticipant* participant,
+        const fastrtps::string_255 topic_name,
+        const fastrtps::string_255 type_name,
+        const fastrtps::types::TypeInformation& type_information)
+{
+    // TODO (check if needed)
+}
+
+void Participant::on_type_discovery(
+        eprosima::fastdds::dds::DomainParticipant* participant,
+        const fastrtps::rtps::SampleIdentity& request_sample_id,
+        const fastrtps::string_255& topic,
+        const fastrtps::types::TypeIdentifier* identifier,
+        const fastrtps::types::TypeObject* object,
+        fastrtps::types::DynamicType_ptr dyn_type)
+{
+    DEBUG("Type discovered: " << dyn_type->get_name() << " in topic: " << topic.to_string());
+
+    // Create TypeSupport and register it
+    eprosima::fastdds::dds::TypeSupport(
+        new eprosima::fastrtps::types::DynamicPubSubType(dyn_type)).register_type(participant_);
+
+    // In case this callback is sent, it means that the type is already registered, so notify
+    // TODO in future it would be better to update every topic in this type name, and not just the one calling here
+    on_topic_discovery_(topic.to_string(), dyn_type->get_name());
+}
+
+////////////////////////////////////////////////////
+// RETRIEVE VALUES METHODS
+////////////////////////////////////////////////////
+
+// TODO erase
+// void Participant::listener(const std::shared_ptr<FastDdsListener>& listener)
+// {
+//     listener_ = listener;
+
+//     // Change listener in every reader handler
+//     for (auto& reader : readers_)
+//     {
+//         reader.second->listener(listener_);
+//     }
+// }
+
+// std::shared_ptr<FastDdsListener> Participant::listener() const
+// {
+//     return listener_;
+// }
+
+////////////////////////////////////////////////////
+// EXTERNAL EVENT METHODS
+////////////////////////////////////////////////////
+
+void Participant::on_topic_discovery_(
+        const std::string& topic_name,
+        const std::string& type_name)
+{
+    // TODO: check if mutex required
+
+    // Check if type is registered in Participant
+    bool is_registered = is_type_registered_(type_name);
+    bool is_already_discovered = false;
+
+    // Check if this topic has already been discovered
+    auto it = discovery_database_->find(topic_name);
+    if (it != discovery_database_->end())
+    {
+        is_already_discovered = true;
+
+        // If the topic is already discovered, check the registered type
+        if (it->second.second)
+        {
+            // The topic had already been discovered and registered, so not sending callback twice
+            logInfo(
+                PLOTJUGGLER_FASTDDS,
+                "Topic " << topic_name << " has already been discovered");
+            return;
+        }
+    }
+
+    if (!is_already_discovered)
+    {
+        // Add topic as discovered
+        discovery_database_->operator [](topic_name) = {type_name, is_registered};
```

Comment:
As already said, revisit the need of having `discovery_database_` as a pointer (moreover, a `shared_ptr`) to avoid the need of these kind of workarounds to do simple things. If seems there is no reason for participants to manage the lifetime of discovery databases. Seems like participants could be initialized with a simpler non-owning reference `DiscoveryDatabase&`

Reply:
As discussed before, I prefer to leave DiscoveryDatabase as shared_ptr so far, as it is used from independent points in the code, and it does not carry any bigger problem than storing a shared_ptr in three points along the code.